File: solutions/r/largest-series-product/largest-series-product.R
# Finds the largest product for a specified number of
# consecutive numbers in a given string of numbers.
largest_series_product <- function(digits, span) {
    stopifnot(span >= 0)
    if (span == 0) {
        return(1)
    }

    (digits <- digits
        |> strsplit("") |> unlist()
        |> as.numeric() |> suppressWarnings())

    stopifnot(span <= length(digits))

    largest <- 0
    for (i in seq(digits)) {
        size <- i + span - 1
        if (size <= length(digits)) {
            largest <- max(largest, prod(digits[i:size]))
        }
    }
    largest
}

File: solutions/lua/largest-series-product/largest-series-product.lua
return function(config)
    assert(config.span >= 0)
    assert(config.span <= #config.digits)
    assert(tonumber(config.digits))

    if config.span == 0 then
        return 1
    end

    local largest = -1
    for i = 0, #config.digits - config.span do
        local product = 1
        for j = i + 1, i + config.span do
            product = product * tonumber(config.digits:sub(j, j))
        end
        largest = math.max(largest, product)
    end
    return largest
end
